rfd/00XY-session-locking.md

```diff
@@ -0,0 +1,250 @@
+Every `SessionLock` create/update/delete should cause an audit event to be
+emitted.
+
+#### Relation to `User.Status`
+
+There already exists a field of `User` resource that is used to
+capture a user lock in connection with failed Web UI login attempts.
+
+This `Status` field and its `LoginStatus` definition are superseded by
+`SessionLock`.  All of its use cases should be converted to `SessionLock`.
+
+The `SessionLock` approach allows to specify locks for entities that are only
+yet to exist or exist merely transiently (such as SSO user objects).  It could
+also help with alleviating the load associated with caching `User` resources.
+
+#### `tctl` support
+
+`SessionLock` resources can be managed using `tctl [get|create|rm]`.  In this
+way, it is possible to update (e.g. delay) or remove a session lock after it
+has been created.
+
+There will be a special `tctl sessions lock` helper provided, to facilitate
+supplying time information when creating new `SessionLock`s, see Scenarios below.
+
+### Disabling new certificates
+
+No new user certificates matching a session lock target should be generated
+while the session lock is in force.  An audit event should be emitted upon such
+an attempt.
+
+A new lock check will be added to the `generateUserCert` function
+in `lib/auth/auth.go`.
+
+### Terminating existing sessions
+
+Terminating an existing session due to a (freshly created) session lock is
+similar to terminating an existing session due to certificate expiry.  The
+support for the latter is covered by `srv.Monitor` and its `Start` routine.
+
+In order to make `srv.Monitor` keep track of all the `SessionLock`s without
+periodically polling the backend, two new fields are added to `srv.MonitorConfig`:
++ `StoredSessionLocks`: a slice of `SessionLock`s known at the time of calling
+  `srv.NewMonitor`;
++ `SessionLockWatcher`: a `types.Watcher` detecting additional puts or deletes
+  of `SessionLock`s pertaining to the current session.
+
+The developed logic should work with all the proxies that already make use of
+`DisconnectExpiredCert`, i.e. SSH, k8s and DB.
+
+### Replicating to trusted clusters
+
+`SessionLock` resources are replicated from the root cluster to leaf clusters
+in a similar manner to how CAs are shared between trusted clusters.
+
+The goal should be achieved by introducing a routine similar to
+`periodicUpdateCertAuthorities`. However instead of polling the backend (with
+the default period of 10 minutes defined in `defaults.LowResPollingPeriod`)
+a `types.Watcher`-based algorithm should be preferred.
+
+To be able to distinguish session locks received from a remote (root) cluster,
+the root cluster should send the `SessionLock` resource a resource label of the
+form:
+```
+teleport.dev/replicated-from: <root-cluster-name>
+```
+
+### Scenarios
+
+#### Creating a permanent lock
+
+```
+$ tctl sessions lock --user=foo@example.com --message="Suspicious activity."
```

Comment:
Nit on the naming, it seems like sessions is really redundant, because it won't only lock the session, but also renewal of the certificate and subsequent logins, right? So could be as well tctl lock --users and tctl rm locks/...